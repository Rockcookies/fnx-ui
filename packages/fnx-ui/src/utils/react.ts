import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
