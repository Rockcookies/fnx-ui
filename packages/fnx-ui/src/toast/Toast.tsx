import React, { forwardRef, ReactNode, useEffect, useState } from 'react';
import useProps from '../hooks/use-props';
import Popup from '../popup';
import { canUseDom } from '../utils/detect';
import { noop } from '../utils/misc';
import { classnames, createBEM } from '../utils/namespace';
import { ToastProps } from './interface';

const NS = 'fnx-toast';
const bem = createBEM(NS);

const BODY_LOCK_CLASS_NAME = `${NS}--unclickable`;

let BODY_CLICK_LOCK_COUNT = 0;

type ToastRequiredProps = Required<
	Pick<
		ToastProps,
		| 'visible'
		| 'duration'
		| 'forbidClick'
		| 'overlay'
		| 'overlayCloseable'
		| 'onClick'
	>
>;

const DEFAULT_PROPS: ToastRequiredProps = {
	visible: false,
	duration: 2000,
	forbidClick: false,
	overlay: false,
	overlayCloseable: false,
	onClick: noop,
};

const Toast = forwardRef<HTMLDivElement, ToastProps>((_props, ref) => {
	const [
		{ visible, duration, forbidClick, overlay, overlayCloseable, onClick },
		{
			mountTo,
			position,
			message,
			icon,
			clickCloseable,
			children,
			className,
			...restProps
		},
	] = useProps<ToastRequiredProps, ToastProps>(DEFAULT_PROPS, _props);

	const [showing, setShowing] = useState(false);

	useEffect(() => {
		setShowing(visible);

		if (visible && duration > 0) {
			const timer = setTimeout(() => {
				setShowing(false);
			}, duration);

			return () => {
				clearTimeout(timer);
			};
		}
	}, [visible, duration]);

	// forbid click
	useEffect(() => {
		if (showing && forbidClick && canUseDom()) {
			if (!BODY_CLICK_LOCK_COUNT) {
				document.body.classList.add(BODY_LOCK_CLASS_NAME);
			}

			BODY_CLICK_LOCK_COUNT++;

			return () => {
				BODY_CLICK_LOCK_COUNT--;

				if (!BODY_CLICK_LOCK_COUNT) {
					document.body.classList.remove(BODY_LOCK_CLASS_NAME);
				}
			};
		}
	}, [showing, forbidClick]);

	const renderIcon = () => {
		const formattedIcon: ReactNode = icon;

		return (
			formattedIcon && <div className={bem('icon')}>{formattedIcon}</div>
		);
	};

	const renderMessage = () => {
		const formattedMessage: ReactNode = message || children;

		return (
			formattedMessage && (
				<div className={bem('message')}>{formattedMessage}</div>
			)
		);
	};

	return (
		<Popup
			visible={showing}
			mountTo={mountTo}
			className={classnames(
				bem([
					position,
					{
						text: !icon,
					},
				]),
				className,
			)}
			renderOnShow={true}
			destroyOnHide={true}
			lockScroll={false}
			overlay={overlay}
			overlayCloseable={overlayCloseable}
			{...restProps}
			ref={ref}
			onClick={(e) => {
				if (clickCloseable) {
					setShowing(false);
				}

				onClick(e);
			}}
		>
			{renderIcon()}
			{renderMessage()}
		</Popup>
	);
});

Toast.displayName = 'Toast';

export default Toast;
