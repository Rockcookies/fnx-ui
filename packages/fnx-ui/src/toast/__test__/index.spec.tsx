import React, { FC } from 'react';
import { act, fireEvent, render, waitFor } from '@testing-library/react';
import Toast from '../index';
import ConfigProvider from '../../config-provider';
import TestsDOM from '../../../test/dom';

describe('<Toast />', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	const ddc = () => document.body.classList;

	it('should close Toast when using closeOnClickOverlay prop and overlay is clicked', () => {
		const onHide = jest.fn();

		const { getByTestId } = render(
			<Toast
				visible={true}
				mountTo={false}
				clickCloseable={true}
				duration={0}
				data-testid="toast"
				onHide={onHide}
			/>,
		);

		jest.runAllTimers();

		fireEvent(
			getByTestId('toast'),
			new MouseEvent('click', {
				bubbles: true,
				cancelable: true,
			}),
		);

		jest.runAllTimers();

		expect(onHide).toHaveBeenCalled();
	});

	it('toast disappeared after duration', async () => {
		const { container } = render(
			<Toast
				visible={true}
				mountTo={false}
				duration={10}
				data-testid="toast"
			/>,
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('create a forbidClick toast', async () => {
		const toast = await waitFor(() =>
			Toast.success({
				id: 'toast',
				forbidClick: true,
				duration: 0,
			}),
		);

		expect(document.getElementById('toast')).toMatchSnapshot();

		expect(ddc().contains('fnx-toast--unclickable')).toBeTruthy();

		await waitFor(() => toast.update({ forbidClick: false }));

		expect(ddc().contains('fnx-toast--unclickable')).not.toBeTruthy();

		await waitFor(() => Toast.clearAll());
		jest.runAllTimers();

		expect(document.getElementById('toast')).toMatchSnapshot();
	});

	it('show toast', async () => {
		await waitFor(() =>
			Toast.show({
				id: 'show-toast',
				message: 'show',
				duration: 0,
			}),
		);

		jest.runAllTimers();
		expect(document.getElementById('show-toast')).toMatchSnapshot();

		await waitFor(() =>
			Toast.success({
				id: 'success-toast',
				message: 'success',
				duration: 0,
			}),
		);

		jest.runAllTimers();
		expect(document.getElementById('success-toast')).toMatchSnapshot();

		await waitFor(() =>
			Toast.fail({
				id: 'fail-toast',
				message: 'fail',
				duration: 0,
			}),
		);

		jest.runAllTimers();
		expect(document.getElementById('fail-toast')).toMatchSnapshot();

		await waitFor(() =>
			Toast.loading({
				id: 'loading-toast',
				message: 'loading',
				duration: 0,
			}),
		);

		jest.runAllTimers();
		expect(document.getElementById('loading-toast')).toMatchSnapshot();
		expect(ddc().contains('fnx-toast--unclickable')).toBeTruthy();

		await waitFor(() => Toast.clearAll());
		jest.runAllTimers();

		expect(document.querySelectorAll('.fnx-toast').length).toBe(0);
		expect(ddc().contains('fnx-toast--unclickable')).not.toBeTruthy();
	});

	it('clear multiple toast', async () => {
		Toast.allowMultiple(true);

		await waitFor(() => {
			Toast.success({ message: '1', duration: 0 });
			Toast.success({ message: '2', duration: 0 });
		});

		jest.runAllTimers();
		expect(document.querySelectorAll('.fnx-toast').length).toBe(2);

		await waitFor(() => Toast.clearAll());
		jest.runAllTimers();

		expect(document.querySelectorAll('.fnx-toast').length).toBe(0);

		Toast.allowMultiple(false);
	});

	it('useToast', async () => {
		const Demo: FC = () => {
			const { success } = Toast.useToast();

			return (
				<div
					className="trigger"
					onClick={() => success('success')}
				></div>
			);
		};
		const element = document.createElement('div');

		const { container } = render(
			<ConfigProvider mountTo={() => element}>
				<Demo />
			</ConfigProvider>,
		);

		const trigger = TestsDOM.mustQuerySelector(container, '.trigger');

		act(() => {
			fireEvent.click(trigger);
		});

		expect(element).toMatchSnapshot();
	});
});
