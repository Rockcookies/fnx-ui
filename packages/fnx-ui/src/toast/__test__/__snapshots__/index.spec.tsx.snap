// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toast /> create a forbidClick toast 1`] = `
<div
  class="fnx-popup fnx-popup--center fnx-toast fnx-popup--fade-enter fnx-popup--fade-enter-active"
  id="toast"
  style="z-index: 2003;"
>
  <div
    class="fnx-toast__icon"
  >
    <span
      class="fnx-icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          d="M122.28096 536.623104c-9.940992-9.925632-11.548672-25.360384-2.78528-36.407296l20.487168-25.828352c8.397824-10.58816 24.108032-13.246464 35.211264-5.835776l177.3312 118.35904c9.353216 6.243328 25.452544 5.430272 34.185216-1.654784l468.5824-380.16c10.532864-8.54528 27.030528-7.817216 36.261888 1.400832l11.542528 11.52512c10.04544 10.03008 9.314304 25.951232-1.215488 36.465664l-502.92736 502.183936c-15.64672 15.624192-41.337856 14.94016-57.445376-1.142784l-219.22816-218.9056z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Toast /> create a forbidClick toast 2`] = `null`;

exports[`<Toast /> show toast 1`] = `
<div
  class="fnx-popup--center fnx-toast fnx-toast--text fnx-popup fnx-popup--fade-enter-done"
  id="show-toast"
  style="z-index: 2004;"
>
  <div
    class="fnx-toast__message"
  >
    show
  </div>
</div>
`;

exports[`<Toast /> show toast 2`] = `
<div
  class="fnx-popup--center fnx-toast fnx-popup fnx-popup--fade-enter-done"
  id="success-toast"
  style="z-index: 2005;"
>
  <div
    class="fnx-toast__icon"
  >
    <span
      class="fnx-icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          d="M122.28096 536.623104c-9.940992-9.925632-11.548672-25.360384-2.78528-36.407296l20.487168-25.828352c8.397824-10.58816 24.108032-13.246464 35.211264-5.835776l177.3312 118.35904c9.353216 6.243328 25.452544 5.430272 34.185216-1.654784l468.5824-380.16c10.532864-8.54528 27.030528-7.817216 36.261888 1.400832l11.542528 11.52512c10.04544 10.03008 9.314304 25.951232-1.215488 36.465664l-502.92736 502.183936c-15.64672 15.624192-41.337856 14.94016-57.445376-1.142784l-219.22816-218.9056z"
        />
      </svg>
    </span>
  </div>
  <div
    class="fnx-toast__message"
  >
    success
  </div>
</div>
`;

exports[`<Toast /> show toast 3`] = `
<div
  class="fnx-popup--center fnx-toast fnx-popup fnx-popup--fade-enter-done"
  id="fail-toast"
  style="z-index: 2006;"
>
  <div
    class="fnx-toast__icon"
  >
    <span
      class="fnx-icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          d="M455.12192 128.99328c-0.187392-8.402944 5.812224-15.215616 14.620672-15.215616h84.514816c8.262656 0 14.81728 6.38976 14.620672 15.21664l-12.035072 538.456064c-0.187392 8.403968-7.324672 15.21664-15.03744 15.21664h-59.611136c-8.117248 0-14.839808-6.38976-15.03744-15.21664L455.12192 128.99328zM512 853.334016c-31.418368 0-56.889344-25.469952-56.889344-56.88832 0-31.419392 25.470976-56.889344 56.889344-56.889344s56.889344 25.469952 56.889344 56.889344c0 31.418368-25.470976 56.88832-56.889344 56.88832z"
        />
      </svg>
    </span>
  </div>
  <div
    class="fnx-toast__message"
  >
    fail
  </div>
</div>
`;

exports[`<Toast /> show toast 4`] = `
<div
  class="fnx-popup--center fnx-toast fnx-popup fnx-popup--fade-enter-done"
  id="loading-toast"
  style="z-index: 2007;"
>
  <div
    class="fnx-toast__icon"
  >
    <span
      class="fnx-icon"
    >
      <svg
        aria-hidden="true"
        class="fnx-icon__spinner"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="25 25 50 50"
        width="1em"
      >
        <circle
          cx="50"
          cy="50"
          fill="none"
          r="20"
        />
      </svg>
    </span>
  </div>
  <div
    class="fnx-toast__message"
  >
    loading
  </div>
</div>
`;

exports[`<Toast /> toast disappeared after duration 1`] = `<div />`;

exports[`<Toast /> useToast 1`] = `
<div>
  <div>
    <div
      class="fnx-popup fnx-popup--center fnx-toast fnx-popup--fade-enter fnx-popup--fade-enter-active"
      style="z-index: 2010;"
    >
      <div
        class="fnx-toast__icon"
      >
        <span
          class="fnx-icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
          >
            <path
              d="M122.28096 536.623104c-9.940992-9.925632-11.548672-25.360384-2.78528-36.407296l20.487168-25.828352c8.397824-10.58816 24.108032-13.246464 35.211264-5.835776l177.3312 118.35904c9.353216 6.243328 25.452544 5.430272 34.185216-1.654784l468.5824-380.16c10.532864-8.54528 27.030528-7.817216 36.261888 1.400832l11.542528 11.52512c10.04544 10.03008 9.314304 25.951232-1.215488 36.465664l-502.92736 502.183936c-15.64672 15.624192-41.337856 14.94016-57.445376-1.142784l-219.22816-218.9056z"
            />
          </svg>
        </span>
      </div>
      <div
        class="fnx-toast__message"
      >
        success
      </div>
    </div>
  </div>
</div>
`;
